File: packages/client-java/src/main/java/com/durablestreams/exception/PreconditionFailedException.java
package com.durablestreams.exception;

import com.durablestreams.model.Offset;

/**
 * Exception thrown when an If-Match precondition fails (412 response).
 *
 * <p>This occurs when using optimistic concurrency control and another writer
 * has modified the stream since the last read.
 */
public class PreconditionFailedException extends DurableStreamException {

    private final String currentETag;
    private final Offset currentOffset;
    private final boolean streamClosed;

    public PreconditionFailedException(String currentETag, Offset currentOffset, boolean streamClosed) {
        super("Precondition failed: stream was modified by another writer", 412);
        this.currentETag = currentETag;
        this.currentOffset = currentOffset;
        this.streamClosed = streamClosed;
    }

    /**
     * Get the current ETag of the stream.
     */
    public String getCurrentETag() {
        return currentETag;
    }

    /**
     * Get the current tail offset of the stream.
     */
    public Offset getCurrentOffset() {
        return currentOffset;
    }

    /**
     * Check if the stream is closed.
     */
    public boolean isStreamClosed() {
        return streamClosed;
    }
}

File: packages/client-php/src/Exception/PreconditionFailedException.php
<?php

declare(strict_types=1);

namespace DurableStreams\Exception;

/**
 * Exception thrown when an If-Match precondition fails (412 response).
 *
 * This occurs when using optimistic concurrency control and another writer
 * has modified the stream since the last read.
 */
class PreconditionFailedException extends DurableStreamException
{
    public function __construct(
        private ?string $currentETag = null,
        private ?string $currentOffset = null,
        private bool $streamClosed = false,
        array $headers = [],
        ?\Throwable $previous = null,
    ) {
        parent::__construct(
            'Precondition failed: stream was modified by another writer',
            'PRECONDITION_FAILED',
            412,
            $headers,
            $previous
        );
    }

    /**
     * Get the current ETag of the stream.
     */
    public function getCurrentETag(): ?string
    {
        return $this->currentETag;
    }

    /**
     * Get the current tail offset of the stream.
     */
    public function getCurrentOffset(): ?string
    {
        return $this->currentOffset;
    }

    /**
     * Check if the stream is closed.
     */
    public function isStreamClosed(): bool
    {
        return $this->streamClosed;
    }
}
